proposals/module-linking/Binary.md

```diff
@@ -0,0 +1,219 @@
+
+## Import Section updates
+
+Updates to
+[`importdesc`](https://webassembly.github.io/spec/core/binary/modules.html#binary-importdesc)
+
+```
+# note that `0x01 0xc0` in MVP-wasm specifies a 1-byte module field of the
+# string 0xc0, but the 0xc0 byte is not valid utf-8, so this was not a valid
+# MVP import
+import ::=
+    ...
+    mod:name  0x01 0xc0 d:importdesc            ->    {module mod, desc d}
+
+importdesc ::=
+    ...
+    0x05 x:typeidx                              ->    module x
+    0x06 x:typeidx                              ->    instance x
+```
+
+The `import` production is modified to allow optionally dropping the `name`
+field. This is done by encoding an invalid utf-8 string so this is an invalid
+MVP import.
+
+**Validation**
+
+* the `func x` production ensures the type `x` is indeed a function type
+* the `module x` production ensures the type `x` is indeed a module type
+* the `instance x` production ensures the type `x` is indeed an instance type
+
+## Module section (100)
+
+A new module section is added
+
+```
+modulesec ::=  t*:section_100(vec(typeidx))           ->        t*
+```
+
+**Validation**
+
+* Each type index `t` points to a module type (e.g. not a function type)
+* This defines the locally defined set of modules, adding to the module index
+  space.
+
+## Instance section (101)
+
+A new section defining local instances
+
+```
+instancesec ::=  i*:section_101(vec(instancedef))     ->    i*
+
+instancedef ::= m:moduleidx e*:vec(exportdesc)        ->    {instantiate m, imports e}
+```
+
+This defines instances in the module, appending to the instance index space.
+Each instance definition declares the module it's instantiating as well as the
+items used to instantiate that instance.
+
+**Validation**
+
+* The type index `m` must point to a module type.
+* The length of `e*` must be the same as the number of imports the module type
+  has
+* The type of each element of `e*` must be a subtype of the type that it's being
+  matched with. Matching happens pairwise with the list of imports on the module
+  type for `m`.
+* Indices of items referred to by `exportdesc` are all in-bounds. Can only refer
+  to imports/previous aliases, since only those sections can precede.
+
+## Alias section (102)
+
+A new module section is added
+
+```
+aliassec ::=  a*:section_102(vec(alias))     ->        a*
+
+alias ::=
+    0x00 i:instanceidx 0x00 e:exportidx      ->       (alias (instance $i) (func $e))
+    0x00 i:instanceidx 0x01 e:exportidx      ->       (alias (instance $i) (table $e))
+    0x00 i:instanceidx 0x02 e:exportidx      ->       (alias (instance $i) (memory $e))
+    0x00 i:instanceidx 0x03 e:exportidx      ->       (alias (instance $i) (global $e))
+    0x00 i:instanceidx 0x05 e:exportidx      ->       (alias (instance $i) (module $e))
+    0x00 i:instanceidx 0x06 e:exportidx      ->       (alias (instance $i) (instance $e))
+    0x01 0x04 m:moduleidx                    ->       (alias parent (module $m))
```

Comment:
Could `(alias parent (module $m))` be given code `0x05` instead of `0x04` for consistency with above and imports/exports?

Reply:
Whoops that's just a typo on my end.